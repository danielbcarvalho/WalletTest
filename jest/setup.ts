import 'react-native-gesture-handler/jestSetup';
import { mockDefaultTheme } from '../src/theme/defaultTheme';

jest.mock('styled-components', () => {
  const original = jest.requireActual('styled-components');

  return {
    ...original,
    useTheme: () => {
      return mockDefaultTheme;
    },
  };
});
