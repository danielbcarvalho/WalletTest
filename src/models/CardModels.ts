export interface Card {
  id: string;
  cvv: string;
  name: string;
  expiry: string;
  number: string;
  kind: 'black' | 'green';
}
