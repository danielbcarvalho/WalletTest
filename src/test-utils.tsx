// TODO: Verificar essa melhoria nos testes
import React from 'react';
import {
  render,
  RenderAPI,
  RenderOptions,
} from '@testing-library/react-native';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import { mockDefaultTheme } from './theme/defaultTheme';

interface IAllThemeProviders {
  children: React.ReactElement;
}
const AllTheProviders = ({ children }: IAllThemeProviders) => {
  return <ThemeProvider theme={mockDefaultTheme}>{children}</ThemeProvider>;
};

function customRender(
  ui: React.ReactElement,
  options = {} as RenderOptions,
): RenderAPI {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export function rendererCreate(
  component: React.ReactElement,
): ReactTestRenderer {
  return renderer.create(<AllTheProviders>{component}</AllTheProviders>);
}

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
