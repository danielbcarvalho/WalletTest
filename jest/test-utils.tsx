import React, { ReactElement, FC } from 'react';
import { render } from '@testing-library/react-native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ThemeProvider } from '../src/theme';

type Options = Parameters<typeof render>[1];

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      cacheTime: Infinity,
    },
  },
});
const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ThemeProvider>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </ThemeProvider>
);

const customRender = (ui: ReactElement, options?: Options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
