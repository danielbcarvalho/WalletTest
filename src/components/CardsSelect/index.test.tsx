import React from 'react';
import { act } from 'react-test-renderer';

import CardsCarousel from '.';
import { fireEvent, render } from '../../../jest/test-utils';
import { Card } from '../../models/CardModels';

describe('CardsCarousel', () => {
  const mockCards: Card[] = [
    {
      id: '4ec42ba9-50af-40d2-af90-8312edbd9ca2',
      number: '3529 5435 3355 8720',
      cvv: '017',
      name: 'John Doe',
      expiry: '12/23',
      kind: 'black',
    },
    {
      name: 'Daniel B Carvalho',
      number: '5203 9436 6133 9001',
      cvv: '123',
      expiry: '11/24',
      id: '4ec42ba9-50af-40d2-af90-8312edbd9XXX',
      kind: 'green',
    },
  ];

  const mockSelectedCardIndex = mockCards.length - 1;

  const mockOnUseCard = jest.fn();

  test('renders correctly with cards', () => {
    const component = render(
      <CardsCarousel cards={mockCards} onUseCard={mockOnUseCard} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders correctly with no cards', () => {
    const component = render(
      <CardsCarousel cards={[]} onUseCard={mockOnUseCard} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('calls onUseCard with correct selected card when "Use Card" button is pressed', async () => {
    const { getByText } = render(
      <CardsCarousel cards={mockCards} onUseCard={mockOnUseCard} />,
    );

    await act(async () => {
      fireEvent.press(getByText('use this card'));
    });

    expect(mockOnUseCard).toHaveBeenCalledWith(
      mockCards[mockSelectedCardIndex],
    );
  });

  test('calls setCardOnTheTop when a card is pressed', async () => {
    const { getByText, toJSON } = render(
      <CardsCarousel cards={mockCards} onUseCard={mockOnUseCard} />,
    );

    await act(async () => {
      fireEvent.press(getByText(mockCards[0].name));
    });

    expect(toJSON()).toMatchSnapshot();
  });
});
