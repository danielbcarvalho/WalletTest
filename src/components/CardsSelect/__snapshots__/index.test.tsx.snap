// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsCarousel calls setCardOnTheTop when a card is pressed 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "width": "100%",
    }
  }
>
  <RCTScrollView
    ListFooterComponent={[Function]}
    bounces={false}
    data={
      [
        {
          "cvv": "123",
          "expiry": "11/24",
          "id": "4ec42ba9-50af-40d2-af90-8312edbd9XXX",
          "kind": "green",
          "name": "Daniel B Carvalho",
          "number": "5203 9436 6133 9001",
        },
        {
          "cvv": "017",
          "expiry": "12/23",
          "id": "4ec42ba9-50af-40d2-af90-8312edbd9ca2",
          "kind": "black",
          "name": "John Doe",
          "number": "3529 5435 3355 8720",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "width": "100%",
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginTop": 0,
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <View
            kind="green"
            style={
              {
                "alignItems": "flex-start",
                "backgroundColor": "#A5FF32",
                "borderBottomLeftRadius": 16,
                "borderBottomRightRadius": 16,
                "borderTopLeftRadius": 16,
                "borderTopRightRadius": 16,
                "display": "flex",
                "gap": 10,
                "height": 180,
                "justifyContent": "space-around",
                "paddingBottom": 20,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 20,
                "width": "100%",
              }
            }
          >
            <Text
              kind="green"
              style={
                {
                  "color": "#3F3F3F",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 18,
                  "lineHeight": 20,
                }
              }
            >
              Green Card
            </Text>
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                kind="green"
                style={
                  {
                    "color": "#3F3F3F",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 16,
                    "lineHeight": 18,
                  }
                }
              >
                Daniel B Carvalho
              </Text>
              <Text
                kind="green"
                style={
                  {
                    "color": "#3F3F3F",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                **** **** **** 9001
              </Text>
              <Text
                kind="green"
                style={
                  {
                    "color": "#3F3F3F",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                validity 11/24
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginTop": -105,
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <View
            kind="black"
            style={
              {
                "alignItems": "flex-start",
                "backgroundColor": "#000000",
                "borderBottomLeftRadius": 16,
                "borderBottomRightRadius": 16,
                "borderTopLeftRadius": 16,
                "borderTopRightRadius": 16,
                "display": "flex",
                "gap": 10,
                "height": 180,
                "justifyContent": "space-around",
                "paddingBottom": 20,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 20,
                "width": "100%",
              }
            }
          >
            <Text
              kind="black"
              style={
                {
                  "color": "#FFFFFF",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 18,
                  "lineHeight": 20,
                }
              }
            >
              Black Card
            </Text>
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                kind="black"
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 16,
                    "lineHeight": 18,
                  }
                }
              >
                John Doe
              </Text>
              <Text
                kind="black"
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                **** **** **** 8720
              </Text>
              <Text
                kind="black"
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                validity 12/23
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFF",
                "fontSize": 16,
                "lineHeight": 28,
                "marginTop": 15,
              }
            }
          >
            use this card
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`CardsCarousel renders correctly with cards 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "width": "100%",
    }
  }
>
  <RCTScrollView
    ListFooterComponent={[Function]}
    bounces={false}
    data={
      [
        {
          "cvv": "017",
          "expiry": "12/23",
          "id": "4ec42ba9-50af-40d2-af90-8312edbd9ca2",
          "kind": "black",
          "name": "John Doe",
          "number": "3529 5435 3355 8720",
        },
        {
          "cvv": "123",
          "expiry": "11/24",
          "id": "4ec42ba9-50af-40d2-af90-8312edbd9XXX",
          "kind": "green",
          "name": "Daniel B Carvalho",
          "number": "5203 9436 6133 9001",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "width": "100%",
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginTop": 0,
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <View
            kind="black"
            style={
              {
                "alignItems": "flex-start",
                "backgroundColor": "#000000",
                "borderBottomLeftRadius": 16,
                "borderBottomRightRadius": 16,
                "borderTopLeftRadius": 16,
                "borderTopRightRadius": 16,
                "display": "flex",
                "gap": 10,
                "height": 180,
                "justifyContent": "space-around",
                "paddingBottom": 20,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 20,
                "width": "100%",
              }
            }
          >
            <Text
              kind="black"
              style={
                {
                  "color": "#FFFFFF",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 18,
                  "lineHeight": 20,
                }
              }
            >
              Black Card
            </Text>
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                kind="black"
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 16,
                    "lineHeight": 18,
                  }
                }
              >
                John Doe
              </Text>
              <Text
                kind="black"
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                **** **** **** 8720
              </Text>
              <Text
                kind="black"
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                validity 12/23
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginTop": -105,
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <View
            kind="green"
            style={
              {
                "alignItems": "flex-start",
                "backgroundColor": "#A5FF32",
                "borderBottomLeftRadius": 16,
                "borderBottomRightRadius": 16,
                "borderTopLeftRadius": 16,
                "borderTopRightRadius": 16,
                "display": "flex",
                "gap": 10,
                "height": 180,
                "justifyContent": "space-around",
                "paddingBottom": 20,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 20,
                "width": "100%",
              }
            }
          >
            <Text
              kind="green"
              style={
                {
                  "color": "#3F3F3F",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 18,
                  "lineHeight": 20,
                }
              }
            >
              Green Card
            </Text>
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                kind="green"
                style={
                  {
                    "color": "#3F3F3F",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 16,
                    "lineHeight": 18,
                  }
                }
              >
                Daniel B Carvalho
              </Text>
              <Text
                kind="green"
                style={
                  {
                    "color": "#3F3F3F",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                **** **** **** 9001
              </Text>
              <Text
                kind="green"
                style={
                  {
                    "color": "#3F3F3F",
                    "fontFamily": "PTSans-Regular",
                    "fontSize": 14,
                    "lineHeight": 16,
                    "textAlign": "center",
                  }
                }
              >
                validity 11/24
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFF",
                "fontSize": 16,
                "lineHeight": 28,
                "marginTop": 15,
              }
            }
          >
            use this card
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`CardsCarousel renders correctly with no cards 1`] = `
<Text
  style={
    {
      "color": "#FFF",
      "fontSize": 16,
      "lineHeight": 28,
      "marginTop": 15,
    }
  }
>
  ops, no card registered
</Text>
`;
